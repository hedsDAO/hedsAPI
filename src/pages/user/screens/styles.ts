import { FlexProps, StackProps, GridProps, GridItemProps, Stack } from '@chakra-ui/react';

/**
 * @constant {FlexProps} $flexStyles
 * @description FlexProps object that contains styles for the Flex component used in
 * the User component for the top half of the page.
 */

export const $flexStyles: FlexProps = {
  justifyContent: { base: 'center', lg: 'space-between' },
  alignItems: { base: 'end', lg: 'baseline' },
  direction: { base: 'column', lg: 'row' },
  maxW: { base: 'full', lg: '3xl', xl: '6xl' },
  mt: 3,
  mx: 'auto',
};

/**
 * @constant {StackProps} $stackStyles
 * @description StackProps object that contains styles for the Stack component used
 * to wrap the user picture and details.
 */

export const $stackStyles: StackProps = {
  alignSelf: 'center',
  alignItems: 'center',
  justifyContent: { base: 'center', lg: 'start' },
  direction: { base: 'column', lg: 'row' },
  gap: { base: 2, lg: 6 },
};

/**
 * @constant {GridProps} $gridStyles
 * @description GridProps object that contains styles for the Grid component used for
 * displaying the user's collection, likes, and discography.
 */

export const $gridStyles: GridProps = {
  mt: { base: 56, lg: 20, xl: 20 },
  maxW: { base: 'full', xl: '6xl' },
  px: { base: 4, md: 10, lg: 20, xl: 0 },
  templateColumns: 'repeat(6, 1fr)',
  mx: 'auto',
  gap: 5,
};

/**
 * @constant {GridItemProps} $spotlightItemStyles
 * @description Spotlight grid item styles
 */

export const $spotlightItemStyles: GridItemProps = {
  rounded: 'lg',
  bg: 'heds.bg3',
  colSpan: { base: 6, xl: 4 },
};

/**
 * @constant {GridItemProps} $recentEventsItemStyles
 * @description Recent Events grid item styles
 */

export const $recentEventsItemStyles: GridItemProps = {
  rounded: 'lg',
  p: 3,
  as: Stack,
  bg: 'heds.bg3',
  colSpan: { base: 6, xl: 2 },
};

/**
 * @constant {GridItemProps} $navItemStyles
 * @description User nav grid item styles
 */

export const $navItemStyles: GridItemProps = {
  colSpan: 6,
  justifyContent: 'space-between',
  alignItems: 'center'
};
