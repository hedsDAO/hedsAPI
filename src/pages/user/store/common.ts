/**
 * @name UserModels
 * @summary Structures used for public user profiles. This can include artists, curators and non-engaged users.
 *
 * @interface User - Base user model; conditionally set
 * @interface ArtistMapping - Mapping for referencing artists by wallet id.
 * @interface TrackMetadataMapping - individual track, sample and submissions within a tape.
 * @interface BadgeData - Structure for user profile badges.
 * @interface UserCollection - conditional interface for user collection object.
 */

export interface User {
  profilePicture: string;
  banner: string;
  twitterHandle: string;
  badges: Array<BadgeData>;
  description: string;
  displayName: string;
  role: number;
  wallet: string;
  samples?: TrackMetadataMapping;
  tracks?: TrackMetadataMapping;
  submissions?: TrackMetadataMapping;
  public: boolean;
  collection?: UserCollection;
}

export interface TrackMetadataMapping {
  [key: string]: {
    [key: string]: {
      [key: string]: TrackMetadata;
    };
  };
}

export interface BadgeData {
  name: string;
  description: string;
  image: string;
}

export interface ArtistMapping {
  [key: string]: User;
}

export interface TrackMetadata {
  audio: string;
  duration: number;
  track: string;
}

export interface UserCollection {
  [key: string]: {
    name: string;
    image: string;
    quantity: number;
  };
}
