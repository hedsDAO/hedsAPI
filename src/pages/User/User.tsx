import { Transition } from '@headlessui/react';
import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import { Dispatch, RootState } from '@/store';
import Container from '@/common/Container';
import { Skeleton } from '@chakra-ui/react';
import { DEFAULT_PROFILE_PICTURE } from '@/common/constants/User';
import { convertSecondsToMinutes, formatWallet } from '@/common/constants/helpers';

export const User = () => {
  const [isCopied, setIsCopied] = useState(false);
  const dispatch = useDispatch<Dispatch>();
  const isUserDataLoading = useSelector((state: RootState) => state.loading.models.userModel);
  const userData = useSelector((state: RootState) => state.userModel);
  const submissions = useSelector((state: RootState) => state.userModel.submissions?.heds?.hedstape);
  const { wallet } = useParams<{ wallet: string }>();
  useEffect(() => {
    if (wallet) dispatch.userModel.getUserData(wallet);
    return () => {
      dispatch.userModel.clearUserState();
    };
  }, []);

  const handleCopy = () => {
    setIsCopied(true);
    const el = document.createElement('textarea');
    el.value = wallet;
    document.body.appendChild(el);
    el.select();
    document.execCommand('copy');
    document.body.removeChild(el);
    const timer = setTimeout(() => {
      setIsCopied(false);
    }, 1000);
    return () => clearTimeout(timer);
  };
  return (
    <Container>
      <div className="aspect-square sm:w-[25%]">
        <img
          src={userData.profilePicture || DEFAULT_PROFILE_PICTURE}
          alt={userData.displayName}
          className="object-cover w-full h-full lg:min-h-[18rem] lg:max-h-[18rem] lg:min-w-[18rem] lg:max-w-[18rem] aspect-square rounded-lg object-center"
        />
        <div className="sm:mt-6 mt-4">
          <h1 className="text-4xl font-semibold tracking-tight text-gray-900 sm:text-3xl">{userData.displayName}</h1>
        </div>
        <p className="text-gray-500 text-sm mb-1 px-1">{userData.description || '...'}</p>
        <Skeleton fadeDuration={2} isLoaded={!isUserDataLoading}>
          <a href={`https://www.twitter.com/${userData?.twitterHandle}`} target="_blank" rel="noreferrer">
            <span className="inline-flex items-center text-xs self-start mt-4 tracking-tight text-neutral-600 px-2.5 py-0.5 rounded-lg transition-all ">
              <i className="fa-brands fa-twitter mr-2 text-[0.75rem]"></i>@{userData.twitterHandle}
            </span>
          </a>
        </Skeleton>
        <Skeleton fadeDuration={3} isLoaded={!isUserDataLoading}>
          <div className="inline-flex">
            <button
              onClick={() => handleCopy()}
              className="inline-flex items-center text-xs self-start mt-4 tracking-tight text-neutral-600  px-2.5 py-0.5 rounded-lg"
            >
              <i className="fa-solid fa-copy mr-2 text-[0.75rem]"></i>
              {userData?.wallet && formatWallet(userData.wallet)}
            </button>
            <Transition
              show={isCopied}
              enter="transition-opacity duration-75"
              enterFrom="opacity-0"
              enterTo="opacity-100"
              leave="transition-opacity duration-150"
              leaveFrom="opacity-100"
              leaveTo="opacity-0"
              className="text-[0.75rem] font-bold mt-auto px-2.5 py-0.5 text-white bg-teal-300 w-24 rounded-lg"
              as="div"
            >
              copied
            </Transition>
          </div>
        </Skeleton>
      </div>
      <div className="sm:w-[75%]">
        <h1 className="text-3xl font-semibold tracking-tight text-gray-900 sm:text-3xl my-2">Submissions</h1>
        <ul role="list" className="divide-y divide-gray-200">
          {submissions &&
            Object.entries(submissions).map(([id, submission], i) => (
              <li className="text-xs text-gray-600" key={i}>
                <a href="#" className="block hover:bg-gray-50">
                  <div className="flex items-center px-4 py-3 sm:px-6 gap-x-2">
                    <span className='text-xs font-thin'>HT{id}</span>
                    <div className="min-w-0 flex-1 sm:flex">{submission.track}</div>
                    <div className="mt-4 flex-shrink-0 sm:mt-0 sm:ml-5">
                      <span className="mr-2">{convertSecondsToMinutes(submission.duration)}</span>
                      <i className="fa-solid fa-play"></i>
                    </div>
                  </div>
                </a>
              </li>
            ))}
        </ul>
        <h1 className="text-3xl font-semibold tracking-tight text-gray-900 sm:text-3xl my-2">Featured On</h1>
      </div>
    </Container>
  );
};
