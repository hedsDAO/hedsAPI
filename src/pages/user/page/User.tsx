import { useSelector } from 'react-redux';
import { RootState, store } from '@/store';
import { TapeData } from '@/models/common';
import { Flex, Image, Stack } from '@chakra-ui/react';
import { UserWrapper, PrivateUserWrapper } from '@/common/wrappers';
import { CopyWalletButton, TwitterLinkButton } from '@/common/buttons';
import { Tracks, Submissions, ProfilePicture, DisplayName, Description, Collection, Badges, Samples } from '@/common/user';
import { isEmpty } from '@/utils';

export const User: React.FC = (): JSX.Element => {
  const loading = useSelector((state: RootState) => state.loading.models.userModel);
  const userData = useSelector((state: RootState) => state.userModel);
  const tapeData = useSelector((state: RootState) => state.tapesModel);
  const userTracks: { [key: string]: TapeData } = store.select.userModel.getTapeCovers(store.getState(), tapeData.allTapes);
  return (
    <div className="bg-[#f5f5f5]">
      <Image shadow={'lg'} src={userData?.banner} className="object-fit object-cover w-screen -mb-32 bg-gray-600 shadow-sm" h="48" />
      <Flex maxWidth={'7xl'} mx={'auto'} flexDirection={['column', 'column', 'row', 'row']} gap={10} px={[10, 4, 2, 0]} py={4}>
        <UserWrapper>
          <Stack direction={'column'}>
            <ProfilePicture loading={loading} userData={userData} />
            <Badges loading={loading} userData={userData} />
            <DisplayName loading={loading} userData={userData} />
            <Description loading={loading} userData={userData} />
            <CopyWalletButton loading={loading} wallet={userData?.wallet} />
            <TwitterLinkButton loading={loading} userData={userData} />
          </Stack>
          <Stack mt={{ base: '0', md: '36' }} direction={'column'} spacing="10" width={'full'}>
            <PrivateUserWrapper loading={loading} userData={userData}>
              <Samples loading={loading} userData={userData} />
              <Tracks loading={loading} userTracks={userTracks} />
              <Submissions loading={loading} userData={userData} />
              <Collection loading={loading} userData={userData} />
            </PrivateUserWrapper>
          </Stack>
        </UserWrapper>
      </Flex>
    </div>
  );
};
