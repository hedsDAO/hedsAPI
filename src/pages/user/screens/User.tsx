import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';

// Components
import { Nav } from '@/components/Nav/Nav';
import { Banner } from '@/pages/user/components/Banner/Banner';
import { Collection } from '@/pages/user/components/Collection/Collection';
import { Details } from '@/pages/user/components/Details/Details';
import { Discography } from '@/pages/user/components/Discography/Discography';
import { Likes } from '@/pages/user/components/Likes/Likes';
import { ProfilePicture } from '@/pages/user/components/ProfilePicture/ProfilePicture';
import { RecentEvents } from '@/pages/user/components/RecentEvents/RecentEvents';
import { Spotlight } from '@/pages/user/components/Spotlight/Spotlight';
import { WalletAndVP } from '@/pages/user/components/WalletAndVP/WalletAndVP';
import { RefreshCollectionButton } from '@/pages/user/components/RefreshCollectionButton/RefreshCollectionButton';
import { UserNavTabs } from '@/pages/user/models/common';
import { Dispatch, store } from '@/store';

// Models & Constants
import { Box, Flex, Grid, GridItem, Stack } from '@chakra-ui/react';
import { USER_NAV_TABS } from '@pages/user/models/constants';
import { Metatags, MetatagTypes } from '@/common/utilities/Metatags';
import * as styles from '@pages/user/screens/styles';

/**
 * @function User
 * @returns {JSX.Element} The `User` component JSX element.
 * @description displays a user's banner, profile picture, details, and
 * various types of content depending on the currently selected tab.
 */

export const User = () => {
  const { wallet } = useParams<{ wallet: string }>();
  const dispatch = useDispatch<Dispatch>();
  const currentTab = useSelector(store.select.navModel.selectCurrentTab);
  const user = useSelector(store.select.userModel.selectUser);

  // Get user data when wallet changes
  useEffect(() => {
    if (wallet?.length) dispatch.userModel.getUser(wallet);
    return () => {
      dispatch.userModel.setSpotlight(null);
    }
  }, [wallet]);

  // Reset pagination when tab changes
  useEffect(() => {
    dispatch.paginationModel.setCurrentPage(0);
    dispatch.paginationModel.setItemsPerPage(4);
  }, [currentTab]);

  return (
    <>
      {user ? (
        <Metatags user={user} type={MetatagTypes.USER}>
          <Box>
            <Banner />
            <Flex {...styles.$flexStyles}>
              <Stack {...styles.$stackStyles}>
                <ProfilePicture />
                <Details />
              </Stack>
              <WalletAndVP />
            </Flex>
            <Grid {...styles.$gridStyles}>
              <GridItem {...styles.$spotlightItemStyles}>
                <Spotlight />
              </GridItem>
              <GridItem {...styles.$recentEventsItemStyles}>
                <RecentEvents />
              </GridItem>
              <GridItem as={Flex} {...styles.$navItemStyles}>
                <Nav tabs={USER_NAV_TABS} />
                <RefreshCollectionButton />
              </GridItem>
              <GridItem {...styles.$navItemStyles}>
                {currentTab === UserNavTabs.COLLECTION && <Collection />}
                {currentTab === UserNavTabs.LIKES && <Likes />}
                {currentTab === UserNavTabs.DISCOGRAPHY && <Discography />}
              </GridItem>
            </Grid>
          </Box>
        </Metatags>
      ) : (
        <></>
      )}
    </>
  );
};
