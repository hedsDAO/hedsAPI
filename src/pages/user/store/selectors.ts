import { RootState } from '@/store';

export const selectUserSubmissions = (state: RootState) => state.userModel.submissions?.heds?.hedstape || {};
export const selectUserBadges = (state: RootState) => state.userModel.badges || [];
export const selectUserTracks = (state: RootState) => state.userModel.tracks?.heds?.hedstape || {};
export const selectUserSamples = (state: RootState) => state.userModel.samples?.heds?.hedstape || {};
export const selectUserDisplayName = (state: RootState) => state.userModel.displayName || '';
export const selectUserBanner = (state: RootState) => state.userModel.banner || '';
export const selectUserWallet = (state: RootState) => state.userModel.wallet || '';
export const selectUserDescription = (state: RootState) => state.userModel.description || '';
export const selectUserTwitterHandle = (state: RootState) => state.userModel.twitterHandle || '';
export const selectUserCollection = (state: RootState) => state.userModel.collection || {};
export const selectUserSubmissionsBySpaceTapeId = (state: RootState, [space, tape, id]: [string, string, string]) =>
  state.userModel?.submissions?.[space]?.[tape]?.[id];
