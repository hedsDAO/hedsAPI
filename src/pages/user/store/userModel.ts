import type { RootModel } from '@/models';
import { Modals } from '@/modules/modals/store/modalModel';
import { createModel } from '@rematch/core';
import { doc, getDoc, setDoc } from 'firebase/firestore';
import { emptyUserState, populateNewUser } from './utils';
import { db } from '@/App';
import { UserRoles, User, HedsTapes } from '@/models/common';
import { formatUserCollection } from '@/utils';
import { Result } from 'ethers/lib/utils';

export interface userModelState {
  connectedUser: User;
  currentUser: User;
}

export const userModel = createModel<RootModel>()({
  state: {} as userModelState,
  reducers: {
    setUserData: (state, payload) => ({ ...state, ...payload }),
    setConnectedUserData: (state, connectedUser: User) => ({ ...state, connectedUser }),
    clearUserState: (state) => emptyUserState(state),
  },
  effects: (dispatch) => ({
    async getConnectedUserData(wallet: string) {
      const docRef = doc(db, 'users', wallet.toLowerCase());
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        this.setConnectedUserData(docSnap.data());
      } else {
        dispatch.modalModel.setModal(Modals.NAME_MODAL);
        dispatch.modalModel.setModalOpen(true);
      }
    },
    async createNewUser([wallet, displayName]: [string, string]) {
      const docRef = doc(db, 'users', wallet.toLowerCase());
      const newUserData = populateNewUser(wallet, displayName);
      await setDoc(docRef, newUserData).then(() => {
        this.setConnectedUserData(newUserData);
      });
    },
    async getUserData(wallet: string) {
      const docRef = doc(db, 'users', wallet.toLowerCase());
      const docSnap = await getDoc(docRef);
      if (docSnap.exists()) {
        this.setUserData(docSnap.data());
      }
    },
    async updateUserCollection([wallet, data, hedsTapes]: [string, Result[], HedsTapes]) {
      const collection = formatUserCollection(data, hedsTapes);
      const docSnap = await getDoc(doc(db, 'users', wallet));
      const userData = docSnap.exists() ? docSnap.data() : null;
      const { role } = userData;
      if (role >= UserRoles.USER) await setDoc(doc(db, 'users', wallet), { ...userData, collection });
      if (role >= UserRoles.ARTIST) await setDoc(doc(db, 'artists', wallet), { ...userData, collection });
      if (role >= UserRoles.CURATOR) await setDoc(doc(db, 'curators', wallet), { ...userData, collection });
      this.setUserData({ ...userData, collection });
    },
  }),
});
