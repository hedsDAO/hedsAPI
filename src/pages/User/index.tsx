import React from 'react';
import { User } from './User';
export default (): JSX.Element => <User />;
