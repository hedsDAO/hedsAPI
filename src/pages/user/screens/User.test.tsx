import { BrowserRouter as Router } from 'react-router-dom';
import { User } from '@/pages/user/screens/User';
import { store } from '@/store';
import { renderWithRematchStore } from '@/tests/renderWithRematchStore';

describe('User unit', () => {
  beforeAll(async () => {});
  describe('snapshot', () => {
    test('snapshot', async () => {
      const snapshot = renderWithRematchStore(
        <Router>
          <User />
        </Router>,
        store,
      );
      expect(snapshot).toMatchSnapshot();
    });
  });
});
