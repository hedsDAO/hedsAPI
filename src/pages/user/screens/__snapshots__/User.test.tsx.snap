// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User unit snapshot snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="css-0"
      >
        <div
          class="chakra-skeleton css-1j9g6ff"
          data-testid="user-banner-skeleton"
        >
          <img
            class="chakra-image css-14uupqj"
            data-testid="user-banner-image"
          />
        </div>
        <div
          class="css-1gwwmb8"
        >
          <div
            class="chakra-stack css-dh93tg"
          >
            <div
              class="chakra-skeleton css-dls4ge"
            >
              <div
                class="css-135glum"
              >
                <img
                  alt="user profile picture"
                  class="chakra-image css-jg1zau"
                />
                <div
                  class="css-1l1z0tr"
                >
                  <p
                    class="chakra-text css-w3qfe8"
                  >
                    Edit Profile
                  </p>
                </div>
              </div>
            </div>
            <div
              class="chakra-stack css-18084o5"
            >
              <div
                class="css-16kpqjv"
              >
                <p
                  class="chakra-text css-t9lnoh"
                />
              </div>
              <div
                class="css-11dwk56"
              >
                <p
                  class="chakra-text css-1itm6b"
                />
              </div>
              <div
                class="css-1yudii9"
              >
                <a
                  class="chakra-link chakra-button css-nsxr0z"
                  data-testid="wallet-button"
                  href="https://www.etherscan.com/address/undefined"
                  target="_blank"
                />
              </div>
            </div>
          </div>
          <div
            class="css-1t562z2"
          >
            <div
              class="chakra-stack css-4vkpib"
            >
              <p
                class="chakra-text gradient-text css-1awbph4"
              >
                92
              </p>
            </div>
            <i
              class="chakra-text fas fa-info-circle css-nw8nzf"
            />
          </div>
        </div>
        <div
          class="css-1qjcsn7"
        >
          <div
            class="css-cys8s9"
          >
            <div
              class="css-1qny5l"
            >
              <div
                class="chakra-stack css-zuobqs"
              >
                <img
                  alt="spotlight cover"
                  class="chakra-image css-1cj8osz"
                />
              </div>
              <div
                class="css-l80tvw"
              >
                <div
                  class="chakra-stack css-5m42i9"
                >
                  <p
                    class="chakra-text css-b3zm5x"
                  >
                    TRACK
                  </p>
                  <p
                    class="chakra-text css-lktipz"
                    data-testid="spotlight-track-name"
                  />
                  <hr
                    aria-orientation="horizontal"
                    class="chakra-divider css-ds3xsz"
                  />
                  <div
                    class="css-1lekzkb"
                  >
                    <div
                      class="chakra-stack css-n21gh5"
                    >
                      <p
                        class="chakra-text css-sqefca"
                      >
                        ARTIST
                      </p>
                      <p
                        class="chakra-text css-jjjdbd"
                        data-testid="spotlight-artist-name"
                      />
                    </div>
                    <div
                      class="css-3xrwiz"
                    >
                      <i
                        class="fa-solid fa-play-circle css-1lkhtdg"
                      />
                      <i
                        class="fa-solid fa-heart css-1dc979u"
                      />
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="css-19s4cpb"
              >
                <div
                  class="css-lxypt5"
                >
                  <i
                    class="fa-solid fa-play-circle css-sylhua"
                  />
                  <i
                    class="fa-solid fa-heart css-18kbln"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="chakra-stack css-1yuqg5l"
          >
            <div
              class="chakra-stack css-n21gh5"
            >
              <div
                class="chakra-stack css-141l2zz"
              />
            </div>
          </div>
          <div
            class="css-6no1tz"
          >
            <div
              class="css-1whh555"
            >
              <button
                class="chakra-button css-wtvw0i"
                type="button"
              >
                Collection
              </button>
              <button
                class="chakra-button css-1fy1hmj"
                type="button"
              >
                Likes
              </button>
              <button
                class="chakra-button css-1fy1hmj"
                type="button"
              >
                Discography
              </button>
            </div>
            <div
              class="css-182nf6q"
            >
              <button
                aria-controls="menu-list-:r1:"
                aria-expanded="false"
                aria-haspopup="menu"
                class="chakra-button chakra-menu__menu-button css-19lqrls"
                id="menu-button-:r1:"
                type="button"
              >
                <span
                  class="css-xl71ch"
                >
                  <svg
                    class="chakra-icon css-e8xa0h"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M 3 5 A 1.0001 1.0001 0 1 0 3 7 L 21 7 A 1.0001 1.0001 0 1 0 21 5 L 3 5 z M 3 11 A 1.0001 1.0001 0 1 0 3 13 L 21 13 A 1.0001 1.0001 0 1 0 21 11 L 3 11 z M 3 17 A 1.0001 1.0001 0 1 0 3 19 L 21 19 A 1.0001 1.0001 0 1 0 21 17 L 3 17 z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
              </button>
              <div
                class="css-0"
                style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
              >
                <div
                  aria-orientation="vertical"
                  class="chakra-menu__menu-list css-miyh87"
                  id="menu-list-:r1:"
                  role="menu"
                  style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);"
                  tabindex="-1"
                >
                  <button
                    class="chakra-menu__menuitem css-3a3z0g"
                    data-index="0"
                    id="menu-list-:r1:-menuitem-:r2:"
                    role="menuitem"
                    tabindex="-1"
                    type="button"
                  >
                    Collection
                  </button>
                  <button
                    class="chakra-menu__menuitem css-1b6qfzf"
                    data-index="1"
                    id="menu-list-:r1:-menuitem-:r3:"
                    role="menuitem"
                    tabindex="-1"
                    type="button"
                  >
                    Likes
                  </button>
                  <button
                    class="chakra-menu__menuitem css-1b6qfzf"
                    data-index="2"
                    id="menu-list-:r1:-menuitem-:r4:"
                    role="menuitem"
                    tabindex="-1"
                    type="button"
                  >
                    Discography
                  </button>
                </div>
              </div>
              <p
                class="chakra-text css-1fj456c"
              >
                Collection
              </p>
            </div>
          </div>
          <div
            class="css-6no1tz"
          >
            <div
              class="css-6no1tz"
            >
              <div
                class="css-yanbl"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-0"
    >
      <div
        class="chakra-skeleton css-1j9g6ff"
        data-testid="user-banner-skeleton"
      >
        <img
          class="chakra-image css-14uupqj"
          data-testid="user-banner-image"
        />
      </div>
      <div
        class="css-1gwwmb8"
      >
        <div
          class="chakra-stack css-dh93tg"
        >
          <div
            class="chakra-skeleton css-dls4ge"
          >
            <div
              class="css-135glum"
            >
              <img
                alt="user profile picture"
                class="chakra-image css-jg1zau"
              />
              <div
                class="css-1l1z0tr"
              >
                <p
                  class="chakra-text css-w3qfe8"
                >
                  Edit Profile
                </p>
              </div>
            </div>
          </div>
          <div
            class="chakra-stack css-18084o5"
          >
            <div
              class="css-16kpqjv"
            >
              <p
                class="chakra-text css-t9lnoh"
              />
            </div>
            <div
              class="css-11dwk56"
            >
              <p
                class="chakra-text css-1itm6b"
              />
            </div>
            <div
              class="css-1yudii9"
            >
              <a
                class="chakra-link chakra-button css-nsxr0z"
                data-testid="wallet-button"
                href="https://www.etherscan.com/address/undefined"
                target="_blank"
              />
            </div>
          </div>
        </div>
        <div
          class="css-1t562z2"
        >
          <div
            class="chakra-stack css-4vkpib"
          >
            <p
              class="chakra-text gradient-text css-1awbph4"
            >
              92
            </p>
          </div>
          <i
            class="chakra-text fas fa-info-circle css-nw8nzf"
          />
        </div>
      </div>
      <div
        class="css-1qjcsn7"
      >
        <div
          class="css-cys8s9"
        >
          <div
            class="css-1qny5l"
          >
            <div
              class="chakra-stack css-zuobqs"
            >
              <img
                alt="spotlight cover"
                class="chakra-image css-1cj8osz"
              />
            </div>
            <div
              class="css-l80tvw"
            >
              <div
                class="chakra-stack css-5m42i9"
              >
                <p
                  class="chakra-text css-b3zm5x"
                >
                  TRACK
                </p>
                <p
                  class="chakra-text css-lktipz"
                  data-testid="spotlight-track-name"
                />
                <hr
                  aria-orientation="horizontal"
                  class="chakra-divider css-ds3xsz"
                />
                <div
                  class="css-1lekzkb"
                >
                  <div
                    class="chakra-stack css-n21gh5"
                  >
                    <p
                      class="chakra-text css-sqefca"
                    >
                      ARTIST
                    </p>
                    <p
                      class="chakra-text css-jjjdbd"
                      data-testid="spotlight-artist-name"
                    />
                  </div>
                  <div
                    class="css-3xrwiz"
                  >
                    <i
                      class="fa-solid fa-play-circle css-1lkhtdg"
                    />
                    <i
                      class="fa-solid fa-heart css-1dc979u"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="css-19s4cpb"
            >
              <div
                class="css-lxypt5"
              >
                <i
                  class="fa-solid fa-play-circle css-sylhua"
                />
                <i
                  class="fa-solid fa-heart css-18kbln"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="chakra-stack css-1yuqg5l"
        >
          <div
            class="chakra-stack css-n21gh5"
          >
            <div
              class="chakra-stack css-141l2zz"
            />
          </div>
        </div>
        <div
          class="css-6no1tz"
        >
          <div
            class="css-1whh555"
          >
            <button
              class="chakra-button css-wtvw0i"
              type="button"
            >
              Collection
            </button>
            <button
              class="chakra-button css-1fy1hmj"
              type="button"
            >
              Likes
            </button>
            <button
              class="chakra-button css-1fy1hmj"
              type="button"
            >
              Discography
            </button>
          </div>
          <div
            class="css-182nf6q"
          >
            <button
              aria-controls="menu-list-:r1:"
              aria-expanded="false"
              aria-haspopup="menu"
              class="chakra-button chakra-menu__menu-button css-19lqrls"
              id="menu-button-:r1:"
              type="button"
            >
              <span
                class="css-xl71ch"
              >
                <svg
                  class="chakra-icon css-e8xa0h"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M 3 5 A 1.0001 1.0001 0 1 0 3 7 L 21 7 A 1.0001 1.0001 0 1 0 21 5 L 3 5 z M 3 11 A 1.0001 1.0001 0 1 0 3 13 L 21 13 A 1.0001 1.0001 0 1 0 21 11 L 3 11 z M 3 17 A 1.0001 1.0001 0 1 0 3 19 L 21 19 A 1.0001 1.0001 0 1 0 21 17 L 3 17 z"
                    fill="currentColor"
                  />
                </svg>
              </span>
            </button>
            <div
              class="css-0"
              style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
            >
              <div
                aria-orientation="vertical"
                class="chakra-menu__menu-list css-miyh87"
                id="menu-list-:r1:"
                role="menu"
                style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);"
                tabindex="-1"
              >
                <button
                  class="chakra-menu__menuitem css-3a3z0g"
                  data-index="0"
                  id="menu-list-:r1:-menuitem-:r2:"
                  role="menuitem"
                  tabindex="-1"
                  type="button"
                >
                  Collection
                </button>
                <button
                  class="chakra-menu__menuitem css-1b6qfzf"
                  data-index="1"
                  id="menu-list-:r1:-menuitem-:r3:"
                  role="menuitem"
                  tabindex="-1"
                  type="button"
                >
                  Likes
                </button>
                <button
                  class="chakra-menu__menuitem css-1b6qfzf"
                  data-index="2"
                  id="menu-list-:r1:-menuitem-:r4:"
                  role="menuitem"
                  tabindex="-1"
                  type="button"
                >
                  Discography
                </button>
              </div>
            </div>
            <p
              class="chakra-text css-1fj456c"
            >
              Collection
            </p>
          </div>
        </div>
        <div
          class="css-6no1tz"
        >
          <div
            class="css-6no1tz"
          >
            <div
              class="css-yanbl"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
